### PR TITLE
drivers/pcie: add support for runtime IRQ detection

### PR DESCRIPTION
**Pull request by ghost**
IRQs assigned to PCI(e) devices, like MMIO windows, are configurable
by firmware and can not necessarily be predicted at build time. This
patch introduces the infrastructure to detect and route PCI IRQs during
driver initialization, if configured and possible.

The NS16550 UART driver is, as always, updated to leverage the new
functionality when PCIe is enabled. The Apollo Lake SoC/UpSquared and
Galileo boards are modified to reflect required configuration changes.

Also, I've taken ownership of the up_squared board.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

### CODE REVIEW

**Comment by gnuless** [comment]
Really need to address the fact that the per-UART boilerplate is 90-odd lines.

**Comment by gnuless** [comment]
Because the per-UART boilerplate has reached 90+ lines, I've made a
template of said boilerplate and use cmake to instantiate instances
at build time. I hate this kind of thing, but it's becoming a real
issue of maintainability (not whizbangneato pointless metaprogramming).

File: CODEOWNERS [hunk 1]
```diff
@@ -89,6 +89,7 @@
 /boards/x86/galileo/                      @nashif
 /boards/x86/quark_d2000_crb/              @nashif
 /boards/x86/quark_se_c1000_devboard/      @nashif
+/boards/x86/up_squared/                   @gnuless
```
Comment (andyross) [inline CODEOWNERS]: Hygiene-wise, probably best to do this as a separate commit.

File: drivers/serial/uart_ns16550_port.tmpl [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+{
+	ARG_UNUSED(dev);
+
+#ifdef UART_NS16550_PCIE_ENABLED
+	if (DEV_CFG(dev)->pcie) {
+		unsigned int irq = DT_UART_NS16550_PORT_@NUM@_IRQ;
+
+		if (irq == PCIE_IRQ_DETECT) {
+			irq = pcie_wired_irq(DT_UART_NS16550_PORT_@NUM@_BASE_ADDR);
+		}
```
Comment (andyross) [inline drivers/serial/uart_ns16550_port.tmpl]: Sort of a metaphor skew going on here: this file is a template designed to automatically generate code correctly given compile-time configuration.  But the test for PCIE_IRQ_DETECT is being done at runtime, when AFAICT there's be no loss of generality to do that statically given that we're doing the same kind of thing already in this very spot.

**Review by andyross — COMMENTED** [review]
Seems like it would be productive to break these changes up into separate commits.  I see a few things happening:

+ CODEOWNERS modification
+ Moving some kconfig settings from board files to the SoC layer
+ Templatizing the 16550's interrupt functions to remove runtime checks
+ Adding support for runtime IRQ probing

Each would be clearer if explained as a single commit.

**Comment by gnuless** [comment]
> Seems like it would be productive to break these changes up into separate commits.

Actually, that's opposite of "productive," as the end result doesn't _produce_ anything more except a longer paper trail.
